api/envoy/service/auth/v3/external_auth.proto

```diff
@@ -103,6 +103,15 @@ message OkHttpResponse {
   // to the downstream client on success. Note that the :ref:`append field in HeaderValueOption <envoy_v3_api_field_config.core.v3.HeaderValueOption.append>`
   // defaults to false when used in this message.
   repeated config.core.v3.HeaderValueOption response_headers_to_add = 6;
+
+  // This field allows the authorization service to set (and overwrite) query
+  // string parameters on the original request before it is sent upstream.
+  repeated config.core.v3.QueryParameter query_parameters_to_set = 7;
+
+  // This field allows the authorization service to specify which query parameters
+  // should be removed from the original request before it is sent upstream. Each
+  // element in this list is a case-sensitive query parameter name to be removed.
```

Comment:
API looks good, but I'd just like to understand some of the implications around case sensitivity. Let's say the authz server is scrubbing something like `some_internal_param=foo`. If an attacker sets `sOme_internal_param=foo`, we don't scrub. Do we worry about backends that interpret query parameters in a case insensitive manner?

To be clear, I'm not asking if query params are case sensitive according to RFC, more so, are there any CVEs or bad backends that are known here? I'm not aware of any myself, but thought it might be worth checking.

Reply:
This is a good question. I think the right way for an authorization service to protect backends in this case is to inspect the entire provided query string (part of the Check request) for parameters to scrub, case-insensitively. Then, for matched keys, either overwrite them (using query parameters to set) or remove them (using query parameters to remove) using the case-sensitive key seen on the original request.

For example, if the backend is known to be case insensitive and depends on a secure QS-param named "some_internal_param" and an attacker sends e.g. Some_internal_param=baz or SOME_INTERNAL_PARAM=bar, the authorization service can strcasecmp(key, "some_Internal_param") and then query_parameters_to_remove.append(key)

Reply:
Yeah, I think this makes a lot of sense to me, thanks.